core/src/main/java/org/apache/iceberg/puffin/PuffinFormat.java

```diff
@@ -0,0 +1,164 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.puffin;
+
+import io.airlift.compress.Compressor;
+import io.airlift.compress.zstd.ZstdCompressor;
+import io.airlift.compress.zstd.ZstdDecompressor;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableTable;
+import org.apache.iceberg.relocated.com.google.common.collect.Table;
+import org.apache.iceberg.util.ByteBuffers;
+
+final class PuffinFormat {
+  private PuffinFormat() {
+  }
+
+  enum Flag {
+    FOOTER_PAYLOAD_COMPRESSED(0, 0),
+    /**/;
+
+    private static final Table<Integer, Integer, Flag> BY_BYTE_AND_BIT = Stream.of(values())
+        .collect(ImmutableTable.toImmutableTable(Flag::byteNumber, Flag::bitNumber, Function.identity()));
+
+    private final int byteNumber;
+    private final int bitNumber;
+
+    Flag(int byteNumber, int bitNumber) {
+      Preconditions.checkArgument(
+          0 <= byteNumber && byteNumber < PuffinFormat.FOOTER_STRUCT_FLAGS_LENGTH,
+          "Invalid byteNumber");
+      Preconditions.checkArgument(0 <= bitNumber && bitNumber < Byte.SIZE, "Invalid bitNumber");
+      this.byteNumber = byteNumber;
+      this.bitNumber = bitNumber;
+    }
+
+    @Nullable
+    static Flag fromBit(int byteNumber, int bitNumber) {
+      return BY_BYTE_AND_BIT.get(byteNumber, bitNumber);
+    }
+
+    public int byteNumber() {
+      return byteNumber;
+    }
+
+    public int bitNumber() {
+      return bitNumber;
+    }
+  }
+
+  static final int FOOTER_START_MAGIC_OFFSET = 0;
+  static final int FOOTER_START_MAGIC_LENGTH = getMagic().length;
+
+  // "Footer struct" denotes the fixed-length portion of the Footer
+  static final int FOOTER_STRUCT_PAYLOAD_SIZE_OFFSET = 0;
+  static final int FOOTER_STRUCT_FLAGS_OFFSET = FOOTER_STRUCT_PAYLOAD_SIZE_OFFSET + 4;
+  static final int FOOTER_STRUCT_FLAGS_LENGTH = 4;
+  static final int FOOTER_STRUCT_MAGIC_OFFSET = FOOTER_STRUCT_FLAGS_OFFSET + FOOTER_STRUCT_FLAGS_LENGTH;
+  static final int FOOTER_STRUCT_LENGTH = FOOTER_STRUCT_MAGIC_OFFSET + getMagic().length;
+
+  static final PuffinCompressionCodec FOOTER_COMPRESSION_CODEC = PuffinCompressionCodec.LZ4;
+
+  static byte[] getMagic() {
```

Comment:
Feels like this would be better as a static member than a method since we're creating a new array each time it's called.

Reply:
it's shared between classes and array isn't constant, so i didn't want to share array constant.

this is called o(1) times overall though, since each class makes its own trusted copy.